src/cast_core/src/SamplingLoader.cpp

```diff
@@ -0,0 +1,154 @@
+// Copyright 2021-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cast_core/actors/SamplingLoader.hpp>
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+
+#include <bsoncxx/builder/basic/document.hpp>
+#include <bsoncxx/builder/basic/kvp.hpp>
+#include <bsoncxx/json.hpp>
+
+#include <mongocxx/client.hpp>
+#include <mongocxx/pool.hpp>
+
+#include <yaml-cpp/yaml.h>
+
+#include <boost/log/trivial.hpp>
+
+#include <gennylib/Cast.hpp>
+#include <gennylib/MongoException.hpp>
+#include <gennylib/context.hpp>
+
+#include <bsoncxx/builder/stream/document.hpp>
+#include <value_generators/DocumentGenerator.hpp>
+
+namespace genny::actor {
+
+/** @private */
+using index_type = std::tuple<DocumentGenerator, std::optional<DocumentGenerator>, std::string>;
+using namespace bsoncxx;
+
+using bsoncxx::builder::basic::kvp;
+using bsoncxx::builder::basic::make_document;
+
+/** @private */
+struct SamplingLoader::PhaseConfig {
+    PhaseConfig(PhaseContext& context,
+                mongocxx::pool::entry& client,
+                uint threadId,
+                size_t totalThreads,
+                ActorId id)
+        : database{(*client)[context["Database"].to<std::string>()]},
+          collection{database[context["Collection"].to<std::string>()]},
+          // The next line uses integer division for non multithreaded configurations.
+          // The Remainder is accounted for below.
+          insertBatchSize{context["InsertBatchSize"].to<IntegerSpec>()},
+          numBatches{context["Batches"].to<IntegerSpec>()},
+          sampleSize{
+              context["SampleSize"].maybe<IntegerSpec>().value_or(numBatches * insertBatchSize)},
+          threadId(threadId) {}
+
+    mongocxx::database database;
+    mongocxx::collection collection;
+    int64_t insertBatchSize;
+    int64_t numBatches;
+    int64_t sampleSize;
+    uint threadId;
+};
+
+void genny::actor::SamplingLoader::run() {
+    for (auto&& config : _loop) {
+        for (auto&& _ : config) {
+            BOOST_LOG_TRIVIAL(info) << "Beginning to run SamplingLoader";
```

Comment:
Is this log useful? I suggest either killing it, making it `debug`, or including more state information about the `config` or something so you can identify further diagnostics when viewing this log.